### ISSUE TITLE
undefined method `publish_timestamp' for nil:NilClass

### ISSUE THREAD

**Issue by shaggyone** [Mjolnir-Software/dynoscale_agent#2]
Hello, I'm experimenting with dynoscale gem and get following error in app response:
```
{
  "status"=>500,
  "title"=>"undefined method `publish_timestamp' for nil:NilClass",
  "meta"=>{
    "exception"=>"undefined method `publish_timestamp' for nil:NilClass",
    "backtrace"=>[
      "/app/vendor/bundle/ruby/3.0.0/gems/dynoscale_ruby-1.0.1/lib/dynoscale_ruby/recorder.rb:38:in `record!'",
      "/app/vendor/bundle/ruby/3.0.0/gems/dynoscale_ruby-1.0.1/lib/dynoscale_ruby/middleware.rb:34:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/activesupport-6.1.3.1/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/middleware/executor.rb:14:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/middleware/static.rb:24:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/rack-2.2.3/lib/rack/sendfile.rb:110:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/middleware/ssl.rb:77:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/actionpack-6.1.3.1/lib/action_dispatch/middleware/host_authorization.rb:92:in `call'",
      "/app/lib/rack_middlewares/jwt_auth.rb:24:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/rack-cors-0.4.1/lib/rack/cors.rb:81:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/railties-6.1.3.1/lib/rails/engine.rb:539:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/puma-3.8.0/lib/puma/configuration.rb:224:in `call'",
      "/app/vendor/bundle/ruby/3.0.0/gems/puma-3.8.0/lib/puma/server.rb:590:in `handle_request'",
      "/app/vendor/bundle/ruby/3.0.0/gems/puma-3.8.0/lib/puma/server.rb:425:in `process_client'",
      "/app/vendor/bundle/ruby/3.0.0/gems/puma-3.8.0/lib/puma/server.rb:289:in `block in run'",
      "/app/vendor/bundle/ruby/3.0.0/gems/puma-3.8.0/lib/puma/thread_pool.rb:120:in `block in spawn_thread'"
    ]
  }
}
```

**Comment by shaggyone** [comment]
We use `puma`, as webserver. It's configured to run in 5 threads. We got 8 such exceptions for 30_000 requests.

**Comment by eabraham** [comment]
Did you notice if these errors occurred during app startup or shutdown? Or were they randomly occurring during the 30_000 request test?

**Comment by shaggyone** [comment]
@eabraham the errors seem to be random.

**Comment by eabraham** [comment]
@shaggyone This is running on a few apps with different versions of Puma without this issue. What version are you running? Could you share your Gemfile with me?

**Comment by leenasn** [comment]
Am also getting this error in the logs once in a while with puma version 6.4.0, Rails 5.2.8. Can you let me know how to fix it? Thanks

**Comment by paulchaplin** [comment]
@eabraham We're also seeing this with Puma, but rather more than 1-in-30k frequency. No apparent correlation with app startup/shutdown, and my gut instinct says it feels like a threading issue.

I saw that you made this change in May 2023: https://github.com/Mjolnir-Software/dynoscale_ruby/commit/f4030c06bdc1a1dc675c156656c780c2d4a2e661

That sounds/looks a _lot_ like it's related to this issue. [Line 38 there](https://github.com/Mjolnir-Software/dynoscale_ruby/commit/f4030c06bdc1a1dc675c156656c780c2d4a2e661#diff-4e74a821a2008082a592844ec8d3d8d5a64692143215c9309913d35e68e2ac45R38) is where this issue's error seems to stem from, when `@@current_report` is `nil`.

I'm not familiar enough with how Dynoscale and Puma interact threads-wise to suggest anything beyond a guess, but assuming your safe-navigation workaround in that commit is harmless, then would something like this also suffice as a change on that line to avoid these errors?

```
if @@current_report
  @@reports[@@current_report.publish_timestamp] = @@current_report
end
```

You could probably also just use safe-nav again if `@@reports` containing `{nil => nil}` doesn't matter:

```
@@reports[@@current_report&.publish_timestamp] = @@current_report
```

**Comment by eabraham** [comment]
Resolved in [v 1.0.4](https://github.com/Mjolnir-Software/delayforreddit/pull/442)
